lib/Doctrine/ODM/MongoDB/UnitOfWork.php

```diff
@@ -919,7 +919,8 @@ private function computeAssociationChanges($parentDocument, $mapping, $value)
                     $this->scheduleCollectionUpdate($value);
                 }
             }
-            $this->visitedCollections[] = $value;
+            // Inline call to UnitOfWork::markCollectionVisited()
+            $this->visitedCollections[spl_object_hash($value)] = $value;
```

Comment:
We could probably benefit from using `spl_object_hash()` in some of these places, although there is a risk of hash conflict. I recall there being old discussions in IRC about using SplObjectStorage for these things, but there was even more overhead. @guilhermeblanco could probably tell you about it.

That said, we probably want to avoid `in_array()` calls in UnitOfWork, and some of these properties should certainly be sets instead of lists. As-is, I'm not sure if there are any precautions against inserting duplicate entries.